File: src/main/java/komu/blunt/stdlib/BasicFunctions.java
package komu.blunt.stdlib;

import com.google.common.base.Objects;
import komu.blunt.objects.EvaluationException;

import java.math.BigInteger;

@SuppressWarnings("unused")
public class BasicFunctions {
    
    @LibraryFunction("primitiveOpPlus")
    @TypeScheme("Num a => (a,a) -> a")
    public static BigInteger plus(BigInteger x, BigInteger y) {
        return x.add(y);
    }
    
    @LibraryFunction("primitiveOpMinus")
    @TypeScheme("Num a => (a,a) -> a")
    public static BigInteger minus(BigInteger x, BigInteger y) {
        return x.subtract(y);
    }

    @LibraryFunction("primitiveOpMultiply")
    @TypeScheme("Num a => (a,a) -> a")
    public static BigInteger multiply(BigInteger x, BigInteger y) {
        return x.multiply(y);
    }

    @LibraryFunction("primitiveOpDivide")
    @TypeScheme("Num a => (a,a) -> a")
    public static BigInteger divide(BigInteger x, BigInteger y) {
        return x.divide(y);
    }

    @LibraryFunction("primitiveMod")
    @TypeScheme("Num a => (a,a) -> a")
    public static BigInteger modulo(BigInteger x, BigInteger y) {
        return x.mod(y);
    }

    @LibraryFunction("primitiveOpLt")
    @TypeScheme("Ord a => (a,a) -> Boolean")
    public static <T extends Comparable<T>> boolean lt(T x, T y) {
        return x.compareTo(y) < 0;
    }

    @LibraryFunction("primitiveOpGt")
    @TypeScheme("Ord a => (a,a) -> Boolean")
    public static <T extends Comparable<T>> boolean gt(T x, T y) {
        return x.compareTo(y) > 0;
    }

    @LibraryFunction("primitiveOpLe")
    @TypeScheme("Ord a => (a,a) -> Boolean")
    public static <T extends Comparable<T>> boolean le(T x, T y) {
        return x.compareTo(y) <= 0;
    }

    @LibraryFunction("primitiveOpGe")
    @TypeScheme("Ord a => (a,a) -> Boolean")
    public static <T extends Comparable<T>> boolean ge(T x, T y) {
        return x.compareTo(y) >= 0;
    }

    @LibraryFunction("primitiveOpEq")
    @TypeScheme("Eq a => (a,a) -> Boolean")
    public static <T> boolean equal(T x, T y) {
        return Objects.equal(x, y);
    }

    @LibraryFunction("unsafe-null")
    public static <T> T unsafeNull() {
        return null;
    }

    @LibraryFunction("fst")
    public static <A, B> A fst(A a, B b) {
        return a;
    }

    @LibraryFunction("snd")
    public static <A, B> B snd(A a, B b) {
        return b;
    }
    
    @LibraryFunction("show")
    public static <T> String show(T o) {
        return String.valueOf(o);
    }

    @LibraryFunction("error")
    public static <T> T error(String message) {
        throw new EvaluationException(message);
    }
}

File: src/main/kotlin/komu/blunt/stdlib/BasicFunctions.kt
package komu.blunt.stdlib

import komu.blunt.eval.RootBindings
import komu.blunt.objects.TypeConstructorValue
import java.math.BigInteger

@Retention(AnnotationRetention.RUNTIME)
annotation class LibraryFunction(val name: String, val scheme: String)

object BasicFunctions {

    fun register(bindings: RootBindings) {
        for (method in javaClass.methods) {
            val libraryFunc = method.getAnnotation(LibraryFunction::class.java)
            if (libraryFunc != null) {
                bindings.bindFunction(libraryFunc.name, libraryFunc.scheme, { s ->
                    if (s is TypeConstructorValue)
                        method.invoke(BasicFunctions, *s.items)
                    else
                        method.invoke(BasicFunctions, s)
                })
            }
        }
    }

    @LibraryFunction("primitiveCompare", "Ord a => (a,a) -> Ordering")
    fun <T : Comparable<T>> primitiveCompare(x: T, y: T): TypeConstructorValue {
        val r = x.compareTo(y)
        return when {
            r < 0 -> BasicValues.LT
            r > 0 -> BasicValues.GT
            else  -> BasicValues.EQ
        }
    }

    @LibraryFunction("show", "a -> String")
    fun show(s: Any): String = s.toString()

    @LibraryFunction("error", "String -> a")
    fun error(s: Any) = throw Exception(s.toString())

    @LibraryFunction("print", "a -> Unit")
    fun primitivePrint(s: Any) = print(s)

    @LibraryFunction("primitiveOpEq", "Eq a => (a,a) -> Boolean")
    fun primitiveOpEq(x: Any, y: Any): TypeConstructorValue = booleanToConstructor(x == y)

    @LibraryFunction("primitiveOpPlus", "Num a => (a,a) -> a")
    fun primitiveOpPlus(x: BigInteger, y: BigInteger): BigInteger = x + y

    @LibraryFunction("primitiveOpMinus", "Num a => (a,a) -> a")
    fun primitiveOpMinus(x: BigInteger, y: BigInteger): BigInteger = x - y

    @LibraryFunction("primitiveOpMultiply", "Num a => (a,a) -> a")
    fun primitiveOpMultiply(x: BigInteger, y: BigInteger): BigInteger = x * y

    @LibraryFunction("primitiveOpDivide", "Num a => (a,a) -> a")
    fun primitiveOpDivide(x: BigInteger, y: BigInteger): BigInteger = x / y

    @LibraryFunction("primitiveMod", "Num a => (a,a) -> a")
    fun primitiveOpMod(x: BigInteger, y: BigInteger): BigInteger = x % y
}
